### DISCUSSION TITLE
Conversion of IOB to spacy JSON taking alot of time ( IOB has 1 million lines)

### DISCUSSION THREAD

**Discussion by Parth-tech** [explosion/spaCy#8490] [Help: Coding & Implementations] [answered]
I just want little guidance that there are 3 IOB file dev, test & train.
**Dev has 1 million lines.
Test has 4 million lines.
Train has 30 million.** 
I am currently just converting dev file as of now because i wasn't sure whether is there any error or not in it.
(the IOB format is correct) It's been **over 3 hours** as of now can idea will this file work or shall I use something else. 

I am **fine-tuning a bert model** using spacy in **google colab** the Runtime hardware chosen is **GPU** and the , and for reference I have followed this article:

https://towardsdatascience.com/how-to-fine-tune-bert-transformer-with-spacy-3-6a90bfe57647

I have followed the exact steps of the article.

I am not familiar with NLP domain neither do I have profound knowledge of pipelining. Can someone please help regarding this, it's really important.

Below i would attach the image regarding time and the statement executed for conversion.
![image](https://user-images.githubusercontent.com/56110372/123139655-11829f00-d474-11eb-897c-450a6a2211e6.png)

**Comment by polm** [comment]
Three hours is a long time, though 1M lines is a lot. 

Couple of things...

1. Do not use screenshots of code, please copy/paste text. 
2. Like the warning says, it is strongly recommended you do not run pip as root. 
3. How long is a "line"?
4. How did you generate this data?

**Reply by Parth-tech** [reply]
Sorry for the screenshot. 

Well actually going through the spacy documentation, there it was mentioned that we can directly convert the file to .spacy binary file.
So the .spacy was generated really quickly.

But I just want to ask you what is your suggestion, what should be done for not running pip as root. I truly do not have any idea about it. Could you guide me on that.

Thanks for the help.

**Reply by polm** [reply]
OK, it sounds like when you converted to `.spacy` directly instead of converting to JSON first it was fast? Glad to hear that solved your problem!

> But I just want to ask you what is your suggestion, what should be done for not running pip as root. I truly do not have any idea about it. Could you guide me on that.

You should use virtualenvs. It may seem annoying and like a lot to learn at first, but it makes using multiple Python environments easy, and is a great way to avoid version related problems. [Here's the homepage](https://virtualenv.pypa.io/en/latest/).

**Reply by Parth-tech** [reply]
Thank you Polm, will look in to the link great help.

**Comment by adrianeboyd** [comment, marked as answer]
This is probably creating one huge document, which is not going to be practical/useful for training or evaluation.

Unless you know that you're only going to be processing single sentences with your model in the future, I'd recommend creating training docs with more than one sentence per document.

As the warning says, you need to include `-s` for it to actually do sentence segmentation during the conversion. Right now `-n 1` is being ignored because there are no boundaries in the input. (This format could include blank lines as sentence boundaries, if you want to provide them from the original data source instead of using the `sentencizer` during the conversion.)

Try:

```bash
spacy convert -n 10 -s -t json file.tsv .
```

If you run into issues with the file size for the training data, be aware that you can provide a directory with multiple JSON or `.spacy` files for training. The `.spacy` format is very compact, so you're less likely to run into problems with it than with JSON.

Also, if you're just starting out / experimenting, it would make sense to start out with just a small subset of your data (something like 100K train and 10K dev tokens), until you're sure that all your processing and training steps are configured as intended, or you might end up spending a lot of time waiting unnecessarily before figuring out that a preprocessing step needs to be modified.

**Reply by Parth-tech** [reply]
Thank you adrianeboyd, well I just directly converted into .spacy and moreover I used -s option.

It was quick and there were no issues while conversion. 

Thank you.
